### DISCUSSION TITLE
Naming conflicts with `windows.h`

### DISCUSSION THREAD

**Discussion by RobLoach** [raysan5/raylib#3945] [General]
Winapi's `windows.h` can have name conflicts with raylib.h . This is a space to talk about workarounds for it. While you are still able to compile raylib on windows, some people still want to use the winapi directly. This is a discussion space to discuss workarounds.

### Name Conflicts
- Rectangle
- LoadImage
- DrawText
- etc

### References
- https://github.com/raysan5/raylib/issues/1217

**Comment by RobLoach** [comment]
fctr recommended the following...
``` c
#define WIN32_LEAN_AND_MEAN
#if defined(_WIN32)
#define WIN32
#endif
#if defined(_WIN64)
#define WIN64
#define _AMD64_
#undef _X86_
#else
#undef _AMD64_
#define _X86_
#endif

#include <minwindef.h>
#include <stdio.h>
#include <stdlib.h>
```

And change

``` c
int main()
```
to
``` c
int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, char* pCmdLine, int nCmdShow)
```

**Reply by ColleagueRiley** [reply]
This only works in some cases though. I don't know what cases that would be though.

**Comment by RobLoach** [comment]
[Availleight](https://github.com/Availleight) had a [C++ solution](https://github.com/raysan5/raylib/issues/1217#issuecomment-1167287717)... 

``` c
namespace raylib{ #include "raylib.h" }
```

And then use...
```c
raylib::BeginDrawing();
raylib::ClearBackground(raylib::Color{255,255,255,255});
raylib::EndDrawing();
```

**Reply by ColleagueRiley** [reply]
This solution just adds a namespace to raylib which it should already have.

**Reply by orcmid** [reply]
@ColleagueRiley 
Namespace facilities are not part of the C Language.  @RobLoach is clear that this is a C++ capability and shows how it gets around certain conflicts without touching `raylib.h`.

But you know that, so I don't understand what you mean by raylib needing to have had it already.  (And you could have said the same thing about `windows.h`.)

**Reply by ColleagueRiley** [reply]
@orcmid  The use of a prefix and the use of a namespace serve a similar purpose. 

Also, winapi is an api designed 38 years ago and is used by every single application with windows support. Raylib on the other hand is not a system api and was designed 10 years ago and has far less of a user base.

To point the finger at winapi for naming conflicts seems a little odd. Although I would agree that winapi should've probably used a prefix, that doesn't justify the lack of prefix for Raylib. 

Not that I want to continue this argument considering Raysan seems set in his design choice. I just wanted to comment on the idea that it's somehow winapi's fault or responsibility.

**Reply by RobLoach** [reply]
Would it be possible to prefix the structs and functions beforehand?
```c
#define Rectangle winapiIsSoOldAndGrossSoMangleIt_Rectangle
#include "windows.h"
#undef Rectangle
```

I have no insight into this though, as I don't use Windows :shrug: ... Optimally though, it would likely be best to use portable libraries.

**Reply by ColleagueRiley** [reply]
> Would it be possible to prefix the structs and functions beforehand?
> 
> ```c
> 
> #define Rectangle winapiIsSoOldAndGrossSoMangleIt_Rectangle
> 
> #include "windows.h"
> 
> #undef Rectangle
> 
> ```
> 
> 
> 
> I have no insight into this though, as I don't use Windows :shrug: ... Optimally though, it would likely be best to use portable libraries.

I will look into this later, thanks.

**Reply by ColleagueRiley** [reply]
> Would it be possible to prefix the structs and functions beforehand?
> 
> ```c
> 
> #define Rectangle winapiIsSoOldAndGrossSoMangleIt_Rectangle
> 
> #include "windows.h"
> 
> #undef Rectangle
> 
> ```
> 
> 
> 
> I have no insight into this though, as I don't use Windows :shrug: ... Optimally though, it would likely be best to use portable libraries.


This actually fixes those naming conflicts (assuming the function isn't being used). thank you so much!

Edit: sorry for the typos.

**Reply by Peter0x44** [reply]
This can help for for some of the problems faced, but not all of them. And it doesn't help if you actually need to call one of the conflicting functions.

I know from some of my experiments with:
https://github.com/raylib-extras/extras-cpp/blob/main/raylib_namespace.h

`namespace raylib{ #include "raylib.h" }`
Firstly, this doesn't compile, # must be on a new line:
```cpp
test.cpp:2:16: error: stray ‘#’ in program
    2 | namespace rl { #include "raylib.h" }
      |                ^
test.cpp:2:17: error: ‘include’ does not name a type
    2 | namespace rl { #include "raylib.h" }
      |                 ^~~~~~~
```
Lets fix that, and try again:
```cpp
#include <windows.h>
namespace rl {
#include "raylib.h"
}

int main(void)
{
        rl::InitWindow(0, 0, 0);
}
```
so far so good... this compiles
but attempting to call a function from windows.h that actually have a conflicting name:
```cpp
#include <windows.h>
namespace rl {
#include "raylib.h"
}

int main(void)
{
        rl::InitWindow(0,0,nullptr);
        CloseWindow(0);
}
```
Results in linker errors:
```
/usr/lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld: /usr/lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/lib/../lib/libuser32.a(libuser32s00081.o):(.text+0x0): multiple definition of `CloseWindow'; /home/peter/pg/raylib/src/libraylib.a(rcore.o):rcore.c:(.text+0x1d3e1): first defined here
collect2: error: ld returned 1 exit status
```

It can help in some circumstances - but not all.
The only sure way will alway be to split up your translation units, and never include raylib.h and windows.h in the same TU.

The primary problem this solves is "I like to prefix library functions". Which is fine, if that's a style goal you have.

**Reply by Autodidac** [reply]
not a bad solution but there is yet a better one, since the namespaces complain in C++20 what you want is to start with the previous.. but to detach the errors all together define them in seperate translation units (files)  namely you want to include raylib in it's own header and only call that from a cpp that includes another with windows. This sufficiently decouples most of the calls that cause the errors in the first place.

**Comment by orcmid** [comment]
> Also, winapi is an api designed 38 years ago and is used by every single application with windows support. Raylib on the other hand is not a system api and was designed 10 years ago and has far less of a user base.

So, 38 years ago would be in the Win16 days of 1986 and in a form unrecognizable now.  Did you miss the part where raylib is a descendant of the [Borland Graphical Interface](https://en.wikipedia.org/wiki/Borland_Graphics_Interface) available for MS-DOS (and under the covers in Turbo Pascal) before Windows was established to any degree.   

I do not suggest that Microsoft should change anything or should have done so.   They have legacy just as raylib does (on a much smaller scale).

**Reply by Peter0x44** [reply]
They keep around legacy that even breaks perfectly normal uses, not just raylib.
To this day, there are "near" and "far" macros that interfere with naming local variables, function arguments, etc.

I'd respect this if were solving a real legacy need, but visual studio today doesn't support any 16 bit system. Why does this need to exist?

A lot of the decisions they make are bizarre and somewhat baseless. You don't need to defend a megacorp worth trillions, it's fine to admit the technical decisions they make are sometimes poor.

I suggest we get to practical discussions of workarounds, which is what this thread is for.

**Comment by Peter0x44** [comment]
If you only have limited surface of the windows API that you want to call, it is very feasible to just expose what you need manually.
For example, if you wish to call MessageBoxA:
```c
(extern "C") __declspec(dllimport) int MessageBoxA(void* hWND, const char* lpText, const char* lpCaption, unsigned int uType);
```
The hardest part is peeling away the ugly macros and typedefs from the declarations, which can be tough, sometimes.
Grep and some careful reading of windows.h can help.

This will also greatly improve compile times, a lone windows.h takes about ~0.9s to compile on my system with WIN32_LEAN_AND_MEAN, without it, it's even worse (~1.6s). Yes, precompiled headers help, but they add more build complexity.

https://nullprogram.com/blog/2023/05/31/
Here is some more information on that specific topic.

**Comment by ColleagueRiley** [comment]
> They keep around legacy that even breaks perfectly normal uses, not just raylib. To this day, there are "near" and "far" macros that interfere with naming local variables, function arguments, etc.
> 
> I'd respect this if were solving a real legacy need, but visual studio today doesn't support any 16 bit system. Why does this need to exist?
> 
> A lot of the decisions they make are bizarre and somewhat baseless. You don't need to defend a megacorp worth trillions, it's fine to admit the technical decisions they make are poor.
> 
> I suggest we get to practical discussions of workarounds, which is what this thread is for.



> They keep around legacy that even breaks perfectly normal uses, not just raylib. To this day, there are "near" and "far" macros that interfere with naming local variables, function arguments, etc.
> 
> I'd respect this if were solving a real legacy need, but visual studio today doesn't support any 16 bit system. Why does this need to exist?
> 
> A lot of the decisions they make are bizarre and somewhat baseless. You don't need to defend a megacorp worth trillions, it's fine to admit the technical decisions they make are poor.
> 
> I suggest we get to practical discussions of workarounds, which is what this thread is for.

1) This thread has been dead for weeks until you sent this 

2) Raylib changing something fundamental would be far less significant. Putting all the blame on winapi would be really silly. Not only that but there are naming issues that exist with all the other system apis.

The "company bad open source project good" argument is silly.

**Reply by RobLoach** [reply]
It is only a problem if you're using winapi, so I'm fine blaming winapi 😉

**Reply by ColleagueRiley** [reply]
> It is only a problem if you're using Winapi, so I'm fine blaming Winapi 😉

That's not true, check rcore_desktop_rgfw.c. It's a problem on every API.

**Reply by Peter0x44** [reply]
Yeah, libX11 has a Font struct that conflicts with raylib's, that's the #2 source of conflicts. It's still not as bad as just dumping every single possible OS functionality you may or may not want in a single enormous header, but it could be better.

Even then you can still encounter some strange ones - I saw someone who created a global variable "select" that caused glfw to crash when trying to make that syscall later.
